promql/engine.go

```diff
@@ -2555,12 +2555,6 @@ func preprocessExprHelper(expr parser.Expr, start, end time.Time) bool {
 }
 
 func newStepInvariantExpr(expr parser.Expr) parser.Expr {
-	if e, ok := expr.(*parser.ParenExpr); ok {
-		// Wrapping the inside of () makes it easy to unwrap the paren later.
-		// But this effectively unwraps the paren.
-		return newStepInvariantExpr(e.Expr)
-
-	}
```

Comment:
The implementation depends on `X` in `(( X ))` be wrapped by the step invariant, and not the `ParenExpr` (see https://github.com/prometheus/prometheus/blob/be2599c853840eb3a86500e2682b8c1f392d4e3d/promql/engine.go#L1177-L1199)

I would see this being done like this inside the `if`:
```go
return &parser.ParenExpr{Expr: newStepInvariantExpr(e.Expr), PosRange: e.PosRange}
```

I think we need unit tests to test the ordering (to also see if it fails if it is otherwise), but that would mostly be out of scope of this PR.

Reply:
Ignore this, wrapping the ParenExpr is the right thing to do

Reply:
That means the code that I pointed out (and other places where these are being unwrapped) needs to be checked and order of unwrap fixed

Reply:
I think I'm still a bit confused here; at this point what is the next step here? It sounds like what it is doing now is correct and the tests are passing. Its unclear to me what you are expecting to get this PR through.

Reply:
Let me try to create some test case which I think will fail. If it does not, then we are all good :)

Reply:
This test case is going to fail, while it should pass:
```diff
diff --git a/promql/engine_test.go b/promql/engine_test.go
index f1761356e..f8e40321e 100644
--- a/promql/engine_test.go
+++ b/promql/engine_test.go
@@ -1081,6 +1081,24 @@ load 1ms
 					Metric: labels.Labels{},
 				},
 			},
+		}, {
+			query: `timestamp(((((metric_timestamp @ 3600)))))`,
+			start: 0, end: 7 * 60, interval: 60,
+			result: Matrix{
+				Series{
+					Points: []Point{
+						{V: 3600, T: 0},
+						{V: 3600, T: 60 * 1000},
+						{V: 3600, T: 2 * 60 * 1000},
+						{V: 3600, T: 3 * 60 * 1000},
+						{V: 3600, T: 4 * 60 * 1000},
+						{V: 3600, T: 5 * 60 * 1000},
+						{V: 3600, T: 6 * 60 * 1000},
+						{V: 3600, T: 7 * 60 * 1000},
+					},
+					Metric: labels.Labels{},
+				},
+			},
 		},
 	}
```

The the fix for that is this
```diff
diff --git a/promql/engine.go b/promql/engine.go
index 4ad32bcef..605a3c3ec 100644
--- a/promql/engine.go
+++ b/promql/engine.go
@@ -1195,8 +1195,8 @@ func (ev *evaluator) eval(expr parser.Expr) (parser.Value, storage.Warnings) {
 			// Matrix evaluation always returns the evaluation time,
 			// so this function needs special handling when given
 			// a vector selector.
-			unwrapParenExpr(&e.Args[0])
 			arg := unwrapStepInvariantExpr(e.Args[0])
+			unwrapParenExpr(&arg)
 			vs, ok := arg.(*parser.VectorSelector)
 			if ok {
 				return ev.rangeEval(nil, func(v []parser.Value, _ [][]EvalSeriesHelper, enh *EvalNodeHelper) (Vector, storage.Warnings) {
```

That is because, earlier AST was like `ParenExpr -> ParenExpr -> StepInvariantExpr`, so we unwrap the `ParenExpr` first and then `StepInvariantExpr`. But now it is changed to `StepInvariantExpr -> ParenExpr -> ParenExpr`, so we need to reverse the unwrapping too.

So I asking to reverse the ordering of unwrapping (you will find 3 occurrences of `unwrapParenExpr` in `engine.go`)

Reply:
That seems insufficient; if I do that change (96d4b2e46) then I get some test failures. The only way to get the tests to pass with the unwrap afterwards is to do that unwrap both before and after (ca3fb8706) which seems clunky but makes some sense (since it would unwrap both in the same method before; it could be in either order).